libraries/lib-module-manager/AsyncPluginValidator.cpp

```diff
@@ -0,0 +1,237 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file AsyncPluginValidation.cpp
+
+  @author Vitaly Sverchinsky
+
+  Part of lib-module-manager library
+
+**********************************************************************/
+
+#include "AsyncPluginValidator.h"
+
+#include <optional>
+#include <mutex>
+
+#include "BasicUI.h"
+#include "IPCChannel.h"
+#include "IPCServer.h"
+#include "spinlock.h"
+#include "XMLTagHandler.h"
+
+#include "PluginIPCUtils.h"
+#include "PluginDescriptor.h"
+#include "PluginHost.h"
+#include "XMLFileReader.h"
+
+AsyncPluginValidator::Delegate::~Delegate() = default;
+
+class AsyncPluginValidator::Impl final : 
+   public IPCChannelStatusCallback,
+   public std::enable_shared_from_this<Impl>
+{
+   //Variables below are accessed from any thread
+
+   IPCChannel* mChannel{nullptr};
+   std::optional<wxString> mRequest;
+
+   //The main reason to use spinlock instead of std::mutex here
+   //is that in most cases there will be no attempts for simultaneous
+   //data access (except OnConnect/Disconnect cases), because of
+   //protocol design
+   spinlock mSync;
+
+   //Variable below is accessed only from the main/UI thread
+
+   Delegate* mDelegate{nullptr};
+   std::unique_ptr<IPCServer> mServer;
+
+   //Variables below is accessed only from worker threads
+
+   details::InputMessageReader mMessageReader;
+
+   //called on main thread only!
+   void StartHost()
+   {
+      auto server = std::make_unique<IPCServer>(*this);
+      if(!PluginHost::Start())
+         mDelegate->OnInternalError("Cannot start plugin host");
+      else
+         mServer = std::move(server);
+   }
+   
+   void HandleInternalError(const wxString& msg) noexcept
+   {
+      try
+      {
+         BasicUI::CallAfter([this, self = weak_from_this(), msg]
+         {
+            if(auto lck = self.lock() && mDelegate != nullptr)
```

Comment:
What you wrote is equivalent to
```
bool lck = (!self.expired() && mDelegate)
```

but does not lock the weak pointer for the intended scope!  It should be
```
if (auto lck = self.lock(); lck && mDelegate)
```

Reply:
I would also write the lambda capture without this, only with self to access this.  And that would have caused a compilation error after the incorrect if, which might make you fix it correctly.

Reply:
Thanks! `this` is needed for class-member access